### PR TITLE
stop condition

### PR DESCRIPTION
**Pull request by mxyng**

### CODE REVIEW

File: api/types.go [hunk 1]
```diff
@@ -124,7 +124,8 @@ type Options struct {
 	MirostatTau      float32 `json:"mirostat_tau,omitempty"`
 	MirostatEta      float32 `json:"mirostat_eta,omitempty"`
 
-	NumThread int `json:"num_thread,omitempty"`
+	NumThread     int    `json:"num_thread,omitempty"`
+	StopCondition []byte `json:"stop_condition,omitempty"`
```
Comment (jmorganca) [inline api/types.go:128]: Very cool - do we want to make this a directive in the `Modelfile`? (maybe that's for a future PR)

Comment (mxyng) [reply api/types.go]: I changed it to string so it shouldn't be a problem anymore

File: llama/llama.go [hunk 2]
```diff
@@ -220,15 +220,25 @@ func (llm *llama) generate(input []C.llama_token, fn func(api.GenerateResponse))
...
-			// call the callback
-			fn(api.GenerateResponse{
-				Response: b.String(),
-			})
-
-			output.PushLeft(token)
-			context.PushLeft(int(token))
-			b.Reset()
+
+		if token == C.llama_token_eos() {
```
Comment (BruceMacD) [inline llama/llama.go:224]: should we check for the llm.StopCondition first in case it is a newline?

File: llama/llama.go [hunk 3]
```diff
@@ -220,15 +220,25 @@ func (llm *llama) generate(input []C.llama_token, fn func(api.GenerateResponse))
...
+			// output is a newline. hold on the token until something else is generated
+		} else if b.String() == llm.StopCondition {
+			// output is the stop condition. stop generation
+			break
+		} else if strings.HasPrefix(llm.StopCondition, b.String()) && b.Len() < len(llm.StopCondition) {
+			// output looks like it may be the stop condition. hold on to the token until we know more
```
Comment (BruceMacD) [inline llama/llama.go:230]: what does holding onto the token mean? It checks the context again on the next token generation?

**Comment by mxyng** [comment]
superseded by #225

**Pull request closed without merging**